### DISCUSSION TITLE
How to deal with “Standard_OStream or Standard_CString”  When I want to export a “BREP” file?

### DISCUSSION THREAD

**Discussion by bluestarrysky2022** [donalffons/opencascade.js#170] [Q&A] [answered]
Hello!
I tried to use "BRepTools.Write_1 ()" to export the brep file, has not been successful.
I find that the parameters "Standard_OStream" and "Standard_CString" are undefined in opencascade.js.
Please help me,Thanks everyone!

**Comment by donalffons** [comment, marked as answer]
I believe the `Standard_OStream` is currently unusable (but if you or someone else has ideas / suggestions on how this could be made to work from JS, please share your ideas). However, the `BRepTools.Write_3` method should work. You should be able to just pass a JS string for the `Standard_CString theFile` argument - the conversion is done automatically.

**Reply by bluestarrysky2022** [reply]
Thanks, I have solved it now, see below:

```yml
additionalCppCode: |
  #include <BRepMesh_IncrementalMesh.hxx>
  #include <BRep_Builder.hxx>
  #include <BRepTools.hxx>
  #include <string>
  #include <iostream>
  #include <sstream>
  
  class OccUtils {
    private:
      OccUtils() {}
    public:
      static std::string ShapeToString(const TopoDS_Shape& shape)
      {
        std::ostringstream oss(std::ios_base::out);
        BRepTools::Write(shape, oss);
        return oss.str();
      }
      static TopoDS_Shape ShapeFromString(const std::string& str)
      {
        TopoDS_Shape shape;
        BRep_Builder brepBuilder;
        std::istringstream iss(str, std::ios_base::in);
        BRepTools::Read(shape, iss, brepBuilder);
        return shape;
      }
  };
```

**Reply by donalffons** [reply]
Great, thanks for sharing!

**Comment by John-fzj** [comment]
We use opencascade.js to write BREP. Now the question is : though **BRepTools.Write_3** return "true", we don't know whether the file "a.brep" is created or not, since we didn't find  "a.brep" in the specified file path. The code is below:
```
var shape = createSquare(this.openCascade, 10, 10, 10, 360, 1);
var brepStr = "a.brep";
var theProgress = new this.openCascade.Message_ProgressRange_1();
var result = this.openCascade.BRepTools.Write_3(shape,brepStr,theProgress)
console.log(result)
```
Thanks for help!

**Comment by bluestarrysky2022** [comment]
Hi,Please look here,Creating Custom Builds | OpenCascade.js (ocjs.org)&nbsp;

additionalCppCode: |   #include <BRepMesh_IncrementalMesh.hxx&gt;   #include <BRep_Builder.hxx&gt;   #include <BRepTools.hxx&gt;   #include <string&gt;   #include <iostream&gt;   #include <sstream&gt;      class OccUtils {     private:       OccUtils() {}     public:       static std::string ShapeToString(const TopoDS_Shape&amp; shape)       {         std::ostringstream oss(std::ios_base::out);         BRepTools::Write(shape, oss);         return oss.str();       }       static TopoDS_Shape ShapeFromString(const std::string&amp; str)       {         TopoDS_Shape shape;         BRep_Builder brepBuilder;         std::istringstream iss(str, std::ios_base::in);         BRepTools::Read(shape, iss, brepBuilder);         return shape;       }   };
&nbsp;You use new opencascade.js to write BREP!
var shape = createSquare(this.openCascade, 10, 10, 10, 360, 1);var str = OccUtils.ShapeToString(shape);console.log(str)
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "donalffons/opencascade.js"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年5月30日(星期二) 下午4:43
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [donalffons/opencascade.js] How to deal with “Standard_OStream or Standard_CString”  When I want to export a “BREP” file? (Discussion #170)





 
We use opencascade.js to write BREP. Now the question is : though BRepTools.Write_3 return "true", we don't know whether the file "a.brep" is created or not, since we didn't find  "a.brep" in the specified file path. The code is below:
 var shape = createSquare(this.openCascade, 10, 10, 10, 360, 1); var brepStr = "a.brep"; var theProgress = new this.openCascade.Message_ProgressRange_1(); var result = this.openCascade.BRepTools.Write_3(shape,brepStr,theProgress) console.log(result)  
Thanks for help!

**Comment by John-fzj** [comment]
We found solution. Maybe js doesn't offer methods to write file to disk, and the file that occ.js created is storaged in occ.js buffer or something else, so we can't find file in our disk. But we can read the file and get the String like below:
```
var theProgress = new this.openCascade.Message_ProgressRange_1();
var result = this.openCascade.BRepTools.Write_3(shape,"test",theProgress)
let brep = this.openCascade.FS.readFile("/" + "test", { encoding:"utf8" });
this.openCascade.FS.unlink("/" + "test");
console.log(result)
console.log(brep)
```
Also, when you send a absolute path to BRepTools.Write_3(), it will return false, since js may not write file to you disk path.

**Comment by bluestarrysky2022** [comment]
Your solution is better than mine. I'll try yours, thank you!






------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "donalffons/opencascade.js"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年5月31日(星期三) 中午1:34
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [donalffons/opencascade.js] How to deal with “Standard_OStream or Standard_CString”  When I want to export a “BREP” file? (Discussion #170)





 
We found solution. Maybe js doesn't offer methods to write file to disk, and the file that occ.js created is storaged in occ.js buffer or something else, so we can't find file in our disk. But we can read the file and get the String like below:
 var theProgress = new this.openCascade.Message_ProgressRange_1(); var result = this.openCascade.BRepTools.Write_3(shape,"test",theProgress) let brep = this.openCascade.FS.readFile("/" + "test", { encoding:"utf8" }); this.openCascade.FS.unlink("/" + "test"); console.log(result) console.log(brep)
